libsolidity/ast/Types.cpp

```diff
@@ -37,13 +37,39 @@
 #include <boost/range/algorithm/copy.hpp>
 #include <boost/range/adaptor/sliced.hpp>
 #include <boost/range/adaptor/transformed.hpp>
+#include <boost/multiprecision/cpp_dec_float.hpp>
 
 #include <limits>
 
 using namespace std;
 using namespace dev;
 using namespace dev::solidity;
 
+namespace
+{
+
+bool fitsPrecision(bigint const _base, bigint const _exp)
+{
+	using boost::multiprecision::log10;
+	using boost::multiprecision::log2;
+	using boost::multiprecision::ceil;
+	using bigfloat = boost::multiprecision::cpp_dec_float_50;
+
+	size_t const bitsMax = 4096;
+	bigint const baseMax = (bigint(2) << bitsMax) - 1;
+	if(_base > baseMax)
+		return false;
+
+	bigfloat digitsNeeded = bigfloat(_exp) * ceil(log10(bigfloat(_base + 1)));
+	bigfloat bitsNeeded = ceil((digitsNeeded - 1) * log2(bigfloat(10)));
```

Comment:
Why calculate ``digitsNeeded`` as an intermediate result?
You might as well just use ``log2`` directly to get ``bitsNeeded`` instead of multiplying with ``log2(10)`` later...

``log2(10) * log10(a) = log2(a)``

I don't think using 10 as base has any advantage for choosing the correct way of rounding things...

What you want to have in the end is: ``2 ^ 4096 - 1 >= _base^_exp`` or, since it's integers:
``2^4096 > _base^_exp``.

Hence you want ``log2(2^4096) > log2(_base^_exp)``.
Hence `` 4096 > _exp * log2(_base)``.

You only need to compensate for potential rounding error in ``log2(_base)``.

Reply:
Although you may want to treat ``_base <= 1`` separately :-).

Reply:
Also: do you only use ``fitsPrecision`` for ``_base >= 0``? You'd probably need to treat negative ``_base`` differently...